clients/include/rocblas_math.hpp

```diff
@@ -24,6 +24,21 @@ inline __host__ float half_to_float(rocblas_half val)
     return _cvtsh_ss(val);
 }
 
+/* ============================================================================================ */
+// Helper function to truncate float to bfloat16
+
+inline __host__ rocblas_bfloat16 float_to_bfloat16_truncate(float val)
+{
+    union
+    {
+        float    fp32;
+        uint32_t int32;
+    } u = {val};
+    rocblas_bfloat16 ret;
+    ret.data = uint16_t(u.int32 >> 16);
+    return ret;
+}
+
```

Comment:
```
inline __host__ rocblas_bfloat16 float_to_bfloat16_truncate(float val)
{
    union
    {
        float    fp32;
        uint32_t int32;
    } u = {val};
    rocblas_bfloat16 ret;
    ret.data = uint16_t(u.int32 >> 16);
    return ret;
}
```


You need to consider the lower bits in the `float` value. The above logic is not exactly correct, although it won't affect the test results at hand. This is correct (also adding `__device__`, to be more universal):
```
inline __host__ __device__ rocblas_bfloat16 float_to_bfloat16_truncate(float val)
{
    union
    {
        float    fp32;
        uint32_t int32;
    } u = {val};
    rocblas_bfloat16 ret;
    ret.data = uint16_t(u.int32 >> 16);
    if(~u.int32 & 0x7f800000 && u.int32 & 0xffff)
         ret.data |= 1; // Preserve signaling NaN
    return ret;
}
```

Reply:
Edit: The last comment had the reverse of the correct condition.

Reply:
You could also use:
```
inline __host__ __device__ rocblas_bfloat16 float_to_bfloat16_truncate(float val)
{
    union
    {
        float    fp32;
        uint32_t int32;
    } u = {val};
    rocblas_bfloat16 ret;
    ret.data = uint16_t(u.int32 >> 16);
    if((u.int32 & 0x7fff0000) == 0x7f800000 && u.int32 & 0xffff)
         ret.data |= 1; // Preserve signaling NaN
    return ret;
}
```
The main thing is that if the exponent bits are all `1`, and the mantissa has any `1`'s in it, then the value is `NaN`, and we need to preserve it. We do so by setting the lowest bit in the result mantissa if any of the 16 lower bits of the `float` were nonzero. We can be conservative, and only do this when the higher bits of the `float` mantissa are all `0`.

If the MSB of the `float` mantissa is `1`, then it's a quiet `NaN` and the lower bits are just payload and don't matter. If the MSB of the `float` mantissa is `0`, then if any of the lower bits are `1`, it indicates signaling `NaN`, and we have to preserve that by setting at least one bit below the MSB in the result mantissa.